src/gfn/env.py

```diff
@@ -227,22 +300,103 @@ def __init__(
         """
-        self.n_actions = n_actions
-        super().__init__(s0, sf, device_str, preprocessor)
-        self.is_discrete = True
+        device = get_device(device_str, default_device=s0.device)
+
+        # The default dummy action is -1.
+        if isinstance(dummy_action, type(None)):
+            dummy_action = torch.tensor([-1], device=device)
+
+        # The default exit action index is the final element of the action space.
+        if isinstance(exit_action, type(None)):
+            exit_action = torch.tensor([n_actions - 1], device=device)
+
+        self.n_actions = n_actions  # Before init, for compatibility with States.
+
+        super().__init__(
+            s0,
+            state_shape,
+            (1,),  # The action_shape is always 1. TODO: is it?
+            dummy_action,
+            exit_action,
+            sf,
+            device_str,
+            preprocessor,
+        )
+
+        self.is_discrete = True  # After init, else it will be overwritten.
+
+    def states_from_tensor(self, tensor: Tensor):
+        """Wraps the supplied Tensor in a States instance & updates masks."""
+        states_instance = self.make_States_class()(tensor)
+        self.update_masks(states_instance)
+        return states_instance
+
+    # In some cases overwritten by the user to support specific use-cases.
+    def reset(
+        self,
+        batch_shape: Optional[Union[int, Tuple[int]]] = None,
+        random: bool = False,
+        sink: bool = False,
+        seed: int = None,
+    ) -> States:
+        """Instantiates a batch of initial states.
+
+        `random` and `sink` cannot be both True. When `random` is `True` and `seed` is
+            not `None`, environment randomization is fixed by the submitted seed for
+            reproducibility.
+        """
+        assert not (random and sink)
+
+        if random and seed is not None:
+            torch.manual_seed(seed)  # TODO: Improve seeding here?
+
+        if batch_shape is None:
+            batch_shape = (1,)
+        if isinstance(batch_shape, int):
+            batch_shape = (batch_shape,)
+        states = self.States.from_batch_shape(
+            batch_shape=batch_shape, random=random, sink=sink
+        )
+        self.update_masks(states)
+
+        return states
+
+    @abstractmethod
+    def update_masks(self, states: type[States]) -> None:
+        """Updates the masks in States.
+
+        Called automatically after each step for discrete environments.
+        """
+
+    def make_States_class(self) -> type[States]:
+        env = self
+
+        class DiscreteEnvStates(DiscreteStates):
+            state_shape = env.state_shape
+            s0 = env.s0
+            sf = env.sf
+            make_random_states_tensor = env.make_random_states_tensor
+            n_actions = env.n_actions
+            device = env.device
+
+        return DiscreteEnvStates
 
     def make_Actions_class(self) -> type[Actions]:
```

Comment:
Same comment here - I know that the class is `Actions` but I still would advocate that this method should be `make_actions_class` to be more inline with PEP 8.